src/data_science/data_science.py

```diff
@@ -0,0 +1,639 @@
+    def get_faults_by_run_id(self, run_id: UUID) -> pd.DataFrame:
+        """Returns a dataframe of all faults of a given run id
+        :param run_id: the run id
+        :return: a dataframe of faults
+        """
+        unix_2020 = 1577836800
+        faults_df = self.log_collector.get_faults(run_id)
+        faults_df["begin_time"] = faults_df["begin_tick"].apply(lambda x: x + unix_2020)
+        faults_df["begin_time"] = faults_df["begin_time"].apply(
+            pd.to_datetime, unit="s"
+        )
+        faults_df["end_time"] = faults_df["end_tick"].apply(lambda x: x + unix_2020)
+        faults_df["end_time"] = faults_df["end_time"].apply(pd.to_datetime, unit="s")
+        faults_df.set_index("begin_time", inplace=True)
+        faults_df["fault_id"] = faults_df["fault_id"].astype("string")
+
+        faults_df["title"] = faults_df["fault_type"].apply(
+            lambda x: " ".join(x.split("_")).capitalize()
+        )
+        return faults_df
+
+    def _get_section_length_by_tick(
+        self,
+        enter_tick: int,
+        leave_tick: int,
+        block_section_length: float,
+        current_tick: int,
+    ) -> float:
+        """Returns the fraction length of a block section that has been passed until a given tick
+        :param enter_tick: enter tick
+        :param leave_tick: leave tick
+        :param block_section_length: block_section_length
+        :param current_tick: current tick
+        :return: block section length
+        """
+        if current_tick > leave_tick:
+            return block_section_length
+        return (
+            block_section_length
+            * (current_tick - enter_tick)
+            / (leave_tick - enter_tick)
+        )
+
+    def _calculate_verkehrsleistung_by_tick(
+        self, block_section_times_df: pd.DataFrame, tick: int, start_tick: int
+    ) -> float:
+        """Returns the verkehrsleistung by a given tick
+        :param block_section_times_df: dataframe of block section times
+        :param tick: tick
+        :param start_tick: start tick
+        :return: verkehrsleistung
+        """
+        source_df = block_section_times_df.drop(
+            block_section_times_df[block_section_times_df["enter_tick"] > tick].index
+        )
+        source_df["dist"] = source_df.apply(
+            lambda row: self._get_section_length_by_tick(
+                row["enter_tick"], row["leave_tick"], row["block_section_length"], tick
+            ),
+            axis=1,
+        )
+        if tick == start_tick:
+            return 0
+        return source_df["dist"].sum() * 3600 / (tick - start_tick)
+
+    def get_verkehrsleistung_time_by_run_id(
+        self, run_id: UUID, delta_tick=10
+    ) -> pd.DataFrame:
+        """Returns the accumulated verkehrsleistung over a period of time by a given run id
+        :param run_id: the run id
+        :param delta_tick: delta tick
+        :return: dataframe of verkehrsleistung
+        """
+        unix_2020 = 1577836800
```

Comment:
Maybe you could make this time variable global in the class as you use it multiple times?

Reply:
Yes, good idea